src/isolate.rs

```diff
@@ -581,37 +602,51 @@ impl Isolate {
     annex_arc
   }
 
-  /// Associate embedder-specific data with the isolate. `slot` has to be
-  /// between 0 and `Isolate::get_number_of_data_slots()`.
-  ///
-  /// 0-indexed slot is used internally by rusty_v8, so users have 3 slots
-  /// left to use.
-  #[inline(always)]
-  pub unsafe fn set_data(&mut self, slot: u32, ptr: *mut c_void) {
-    assert!(slot < 4);
-    v8__Isolate__SetData(self, slot + Self::INTERNAL_SLOT_COUNT, ptr)
-  }
-
   /// Retrieve embedder-specific data from the isolate.
   /// Returns NULL if SetData has never been called for the given `slot`.
   pub fn get_data(&self, slot: u32) -> *mut c_void {
-    assert!(slot < 4);
-    unsafe { v8__Isolate__GetData(self, slot + Self::INTERNAL_SLOT_COUNT) }
+    self.get_data_internal(Self::INTERNAL_DATA_SLOT_COUNT + slot)
+  }
+
+  /// Associate embedder-specific data with the isolate. `slot` has to be
+  /// between 0 and `Isolate::get_number_of_data_slots()`.
+  #[inline(always)]
+  pub fn set_data(&mut self, slot: u32, data: *mut c_void) {
+    self.set_data_internal(Self::INTERNAL_DATA_SLOT_COUNT + slot, data)
   }
 
   /// Returns the maximum number of available embedder data slots. Valid slots
-  /// are in the range of 0 - `Isolate::get_number_of_data_slots() - 1`.
+  /// are in the range of `0 <= n < Isolate::get_number_of_data_slots()`.
   pub fn get_number_of_data_slots(&self) -> u32 {
-    unsafe {
-      v8__Isolate__GetNumberOfDataSlots(self) - Self::INTERNAL_SLOT_COUNT
-    }
+    Self::EMBEDDER_DATA_SLOT_COUNT - Self::INTERNAL_DATA_SLOT_COUNT
+  }
+
+  #[inline(always)]
+  pub(crate) fn get_data_internal(&self, slot: u32) -> *mut c_void {
+    let slots = unsafe {
+      let p = self as *const Self as *const u8;
+      let p = p.add(Self::EMBEDDER_DATA_OFFSET);
+      let p = p as *const [*mut c_void; Self::EMBEDDER_DATA_SLOT_COUNT as _];
+      &*p
+    };
```

Comment:
the Isolate ptr won't move right? If so, can this block be pre-computed and stored?

Reply:
> the Isolate ptr won't move right? If so, can this block be pre-computed and stored?

I'm not sure what you mean.

Reply:
I'm on mobile, I mean something along the lines of:

```rust
strict Isolate {
  cxx_isolate: *mut c_void,
  slots: *mut [*mut c_void; SLOT_COUNT]
}
```

I don't think it'll have a significant impact on perf so feel free to ignore this.